workflows4s-core/src/main/scala/workflows4s/mermaid/MermaidRenderer.scala

```diff
@@ -0,0 +1,177 @@
+        } yield stepId.some
+    }
+
+  private val eventShape     = "stadium"
+  private val conditionShape = "hex"
+
+  private def cleanActiveNodes: State[RenderState, Seq[ActiveNode]]            = State { s => s.copy(activeNodes = Seq()) -> s.activeNodes }
+  private def cleanPendingErrors: State[RenderState, Seq[PendingError]]        = State { s => s.copy(pendingErrors = Seq()) -> s.pendingErrors }
+  private def setActiveNodes(nodes: Seq[ActiveNode]): State[RenderState, Unit] = State.modify(_.copy(activeNodes = nodes))
+  private def setNextLinkLabel(label: NextLinkLabel): State[RenderState, Unit] = State.modify { s =>
+    s.copy(activeNodes = s.activeNodes.map(_._1 -> label))
+  }
+
+  private def addNode(nodeF: NodeId => MermaidElement, active: Boolean, label: Option[String] = None): State[RenderState, NodeId] = State { s =>
+    val nodeId = s"node${s.idIdx}"
+    s.copy(
+      chart = s.chart.addElement(nodeF(nodeId)),
+      idIdx = s.idIdx + 1,
+      activeNodes = if (active) s.activeNodes.appended(nodeId -> label) else s.activeNodes,
+    ) -> nodeId
+  }
+  private def addStep(label: String, shape: Option[String] = None): State[RenderState, NodeId]                                    = {
+    addStepGeneral(id => Node(id, label, shape))
+  }
+  private def addStepGeneral(createElem: NodeId => MermaidElement): State[RenderState, NodeId]                                    = {
+    for {
+      prev <- cleanActiveNodes
+      id   <- addNode(createElem, active = true)
+      _    <- addLinks(prev, id)
+    } yield id
+  }
+
+  private def addPendingError(from: NodeId, err: WIOModel.Error): State[RenderState, Unit] =
+    State.modify(_.addPendingError(from, err))
+
+  private def handleErrors(errors: Seq[(NodeId, WIOModel.Error)], to: NodeId): State[RenderState, Unit] = State.modify { state =>
+    val links = errors.map(pendingErr => Link(pendingErr._1, to, s"fa:fa-bolt ${pendingErr._2.name}".some, midfix = "."))
+    state.addElements(links)
+  }
+  private def addLink(from: NodeId, to: NodeId, label: Option[String] = None): State[RenderState, Unit] =
+    addLinks(Seq((from, label)), to)
+  private def addLinks(from: Seq[(NodeId, NextLinkLabel)], to: NodeId): State[RenderState, Unit]        =
+    State.modify(_.addElements(from.map(f => Link(f._1, to, f._2))))
+
+  private def addSubgraph[T](subgraph: State[RenderState, T]): State[RenderState, (Seq[ActiveNode], T)] = State { state =>
+    val id                  = s"node${state.idIdx}"
+    val (subState, subProc) = subgraph.run(RenderState.initial(state.idIdx + 1).copy(activeNodes = state.activeNodes)).value
+    (
+      state.copy(
+        chart = state.chart.addElement(Subgraph(id, " ", subState.chart.elements)),
+        idIdx = state.idIdx + subState.idIdx + 1,
+        activeNodes = Seq((id, None)),
+        pendingErrors = state.pendingErrors ++ subState.pendingErrors,
+      ),
+      (subState.activeNodes, subProc),
+    )
+  }
+
+  private type NodeId        = String
```

Comment:
maybe `opaque type`?

Reply:
I dont think its needed right now. Im all for it if the type leaves the class scope. But for relatively small scope type alias should be enough.

Reply:
sounds good